File: Udemy/Practicas/CursoDePython_Udemy/Ejercicios/Calculadora.py
# Prgrama de calculadora básica para consola en Python
# Autor: David Rodríguez

# Creamos una función suma() con el parámetro cantidad
def suma(cantidad):
    suma = None
    
    # Iteramos con un for la cantidad de números a sumar y calculamos su resultado
    for i in range(cantidad):
        i += 1
        numero = eval(input(f"Digita el número {i}: ")) # Pedimos un número al usuario
        
        # Dado el caso que suma sea igual a None le agregaremos lo que tenemos en numero
        if suma == None:
            suma = numero
        else:
            suma = suma + numero # Si suma ya no es igual None, haremos la respectiva operación

    print(f"\nEl resultado de la suma es: {suma}\n") # Imprimimos el resultado de la suma

# Creamos una función resta() con el parámetro cantidad
def resta(cantidad):
    resta = None
    
    # Iteramos con un for la cantidad de números a restar y calculamos su resultado
    for i in range(cantidad):
        i += 1
        numero = eval(input(f"Digita el número {i}: ")) # Pedimos un número al usuario
        
        # Dado el caso que resta sea igual a None le agregaremos lo que tenemos en numero
        if resta == None:
            resta = numero
        else:
            resta = resta - numero # Si resta ya no es igual None, haremos la respectiva operación

    print(f"\nEl resultado de la resta es: {resta}\n") # Imprimimos el resultado de la resta

# Creamos una función multiplicacion() con el parámetro cantidad
def multiplicacion(cantidad):
    multiplicacion = None
    
    # Iteramos con un for la cantidad de números a multiplicar y calculamos su resultado
    for i in range(cantidad):
        i += 1
        numero = eval(input(f"Digita el número {i}: ")) # Pedimos un número al usuario
        
        # Dado el caso que multiplicacion sea igual a None le agregaremos lo que tenemos en numero
        if multiplicacion == None:
            multiplicacion = numero
        else:
            multiplicacion = multiplicacion * numero # Si multiplicacion ya no es igual None, haremos la respectiva operación

    print(f"\nEl resultado de la multiplicación es: {multiplicacion}\n") # Imprimimos el resultado de la multiplicación

# Creamos una función division() con el parámetro cantidad
def division(cantidad):
    division = None
    
    # Iteramos con un for la cantidad de números a multiplicar y calculamos su resultado
    for i in range(cantidad):
        i += 1
        numero = eval(input(f"Digita el número {i}: ")) # Pedimos un número al usuario
        
        # Dado el caso que division sea igual a None le agregaremos lo que tenemos en numero
        if division == None:
            division = numero
        else:
            division = division / numero # Si division ya no es igual None, haremos la respectiva operación

    print(f"\nEl resultado de la división es: {division}\n") # Imprimimos el resultado de la división

# Creamos una función resto() con el parámetro cantidad
# El usuario solo ṕodrá calcular la cantidad definida por el sistema
# Es decir solo podrá calcular 2 números
def resto(cantidad):
    resto = None
    print("Solo puedes calcular dos números\n")
    
    # Iteramos con un for la cantidad de números a multiplicar y calculamos su resultado
    for i in range(cantidad):
        i += 1
        numero = eval(input(f"Digita el número {i}: ")) # Pedimos un número al usuario
        
        # Dado el caso que resto sea igual a None le agregaremos lo que tenemos en numero
        if resto == None:
            resto = numero
        else:
            resto = resto % numero # Si resto ya no es igual None, haremos la respectiva operación

    print(f"\nEl resultado del resto de la división es: {resto}\n") # Imprimimos el resultado

# Creamos una función para el menú de opciones
def opciones():
    nombre = input("Bienvenido..."
                   "\nDigita tu nombre: ") # Pedimos el nombre del usuario
    
    # Iteramos el menú de opciones hasta que el usuario rompa el ciclo
    while(True):
        opciones = eval(input(f"{nombre} escoge una opción:\n"
                            "\n1.) Suma."
                            "\n2.) Resta."
                            "\n3.) Multiplicación."
                            "\n4.) División."
                            "\n5.) Resto de una división."
                            "\n0.) Salir"
                            "\n\nOpción: "))
        
        # Dado el caso de que el usuario escoja la opción 1
        if opciones == 1:
            print("\nHaz elegido la opción: \"1.) Suma.\"")
            cantidad = eval(input("Digita una cantidad de números a sumar: ")) # Le pedimos al usuario que digite una cantidad
            print()

            suma(cantidad) # Enviamos esa cantidad a nuestra función de suma()
            
        # Dado el caso de que el usuario escoja la opción 2
        elif opciones == 2:
            print("\nHaz elegido la opción: \"1.) Resta.\"")
            cantidad = eval(input("Digita una cantidad de números a restar: ")) # Le pedimos al usuario que digite una cantidad
            print()

            resta(cantidad) # Enviamos esa cantidad a nuestra función de resta()

        # Dado el caso de que el usuario escoja la opción 3
        elif opciones == 3:
            print("\nHaz elegido la opción: \"1.) Multiplicación.\"")
            cantidad = eval(input("Digita una cantidad de números a multiplicar: ")) # Le pedimos al usuario que digite una cantidad
            print()

            multiplicacion(cantidad) # Enviamos esa cantidad a nuestra función de multiplicacion()

        # Dado el caso de que el usuario escoja la opción 4
        elif opciones == 4:
            print("\nHaz elegido la opción: \"1.) División.\"")
            cantidad = eval(input("Digita una cantidad de números a dividir: ")) # Le pedimos al usuario que digite una cantidad
            print()

            division(cantidad) # Enviamos esa cantidad a nuestra función de division()

        # Dado el caso de que el usuario escoja la opción 5
        elif opciones == 5:
            print("\nHaz elegido la opción: \"1.) Resto de una división.\"")
            cantidad = 2 # Para este caso, el número 2 será la cantidad definida por el sistema
            resto(cantidad) # Enviamos esa cantidad a nuestra función de resto()

        # Dado el caso de que el usuario escoja la opción 0
        elif opciones == 0:
            print("\nHaz elegido la opción: \"0.) Salir.\""
                  "\n\nHasta pronto!!!...")
            break # EL programa romperá el ciclo y dará fin a su ejecución
        
        # Dado el caso de que el usuario digite una opción incorrecta
        else:
            print("\nError!!!... Opción incorrecta.\n")

# Función principal de ejecució del programa
def main():
    opciones() # Llamamos a la función de opciones()
    
if __name__ == "__main__":
    main()

File: OpenBootcamp/Desarrollo_Mobile/Practicas/Pruebas/calculadora/bin/calculadora.dart
// Calculadora en consola de Dart
// Autor: David Rodríguez

import 'dart:io'; // Imprtamos la librería "io"

void main(List<String> arguments) {
  String ? nombre; // Declaramos una variable "nombre"

  // Generamos un mensaje de bienvenida y pedimos que se ingrese un nombre
  print("Bienvenido!!!... Digita tu nombre: ");
  nombre = stdin.readLineSync()!;
  
  // Llamamos a la función "menuOpciones()" y le pasamos como parámetro el nombre del usuario
  menuOpciones(nombre);
}

void menuOpciones(String nombre) {
  // Variables a usar dentro de la función
  int ? opcion;
  int ? cantidad;
  double ? numero;
  double ? suma;
  double ? resta;
  double ? multiplicacion;
  double ? division;
  double ? porcentaje;
  bool pass = false; // Usamos esta variable como condicional para la ejecución del programa

  // Se ejcutará lo diguiente mientras "pass" sea igual a false
  while (pass == false) {
      opcion = getInt("\n$nombre digita una opción\n"
                      "\n1.) Suma."
                      "\n2.) Resta."
                      "\n3.) Multiplicación."
                      "\n4.) División."
                      "\n5.) Porcentaje de un número."
                      "\n0.) Salir."
                      "\n\nOpción:"); // Pedimos que se ingrese una opción("\n$nombre digita una opcion:\n"

    // Si no se presiona ctrl + D, el programa continuará con su ejecución
    switch (opcion) {
      case 1:
        // Si la opción ingresada es 1
        cantidad = getInt("\nDigita una cantidad de números a sumar:"); // Pedimos que se ingrese una cantidad de números
        suma = sumaOperacion(cantidad); // Enviamos dicha cantidad a "sumaOperacion()"
        print("El resultado de la suma es: $suma"); // Imprimimos el resultado de la suma
        break;
      case 2:
        // Si la opción ingresada es 2
        cantidad = getInt("\nDigita una cantidad de números a restar:"); // Pedimos que se ingrese una cantidad de números
        resta = restaOperacion(cantidad); // Enviamos dicha cantidad a "restaOperacion()"
        print("El resultado de la resta es: $resta"); // Imprimimos el resultado de la resta
        break;
      case 3:
        // Si la opción ingresada es 3
        cantidad = getInt("\nDigita una cantidad de números a sumar:"); // Pedimos que se ingrese una cantidad de números
        multiplicacion = multiplicacionOperacion(cantidad); // Enviamos dicha cantidad a "multiplicacionOperacion"
        print("El resultado de la multiplicación es: $multiplicacion"); // Imprimimos el resultado de la multiplicación
        break;
      case 4:
        // Si la opción ingresada es 4
        cantidad = getInt("\nDigita una cantidad de números a dividir:"); // Pedimos que se ingrese una cantidad de números
        division = divisionOperacion(cantidad); // Enviamos dicha cantidad a "divisionOperacion()"

        if (division == null) {
          print("Error!!!... División entre cero"); // Si "division" es igual a null, se nos muestra este error
        } else {
          print("El resultado de la división es: ${division.toStringAsFixed(3)}"); // Si no, entonces imprimimos el resultado redondeado a 3 decimales
        }

        break;
      case 5:
        // Si la opción ingresada es 5
        porcentaje = porcentajeOperacion(); // Instanciamos a la función "porcentajeOperacion()"
        print("El porcentaje de $numero es: ${porcentaje.toStringAsFixed(2)}"); // Imprimimos el porcentaje redondeado a 3 decimales
        break;
      case 0:
        // Si la opción ingresada es 0
        print("\nHasta luego..."); // Imprimimos un mensaje de despedida
        pass = true; // "pass" será igual a true y se cerrará la ejecución del programa
        break;
      default:
        print("Error!!!... Opción incorrecta"); // Imprimimos un error si la opción ingresada es incorrecta
        break;
    }
  }
}

// Creamos una función "sumaOperacion()" y le psamos como prámetro una cantidad de números
double ? sumaOperacion(int cantidad) {
  double ? numero; // Declaramos una variable "numero"
  double ? suma; // Declaramos una variable "suma"
  List listaNumeros = []; // Declaramos una lista vacia de números
  
  print(""); // Imprimimos un salto de línea

  // Haremos una iteración hasta llegar a "cantidad"
  for (int i = 0; i < cantidad; i++) {
    numero = getDouble("Digita el número ${i + 1}:"); // En cada iteración pediremos un número y lo enviaremos a "getDouble()"
    listaNumeros.add(numero); // Agregaremos el número ingresado a la lista de números
  }

  print("\nLista de números sumados: $listaNumeros\n"); // Imprimimos la lista de números

  // A continuación haremos la suma de cada número de la lista
  // Recorremos todos los elementos de la lista
  for (double num in listaNumeros) {
    if (suma == null) {
      suma = num; // Si "suma" es igual a null le agregamos el primer elemento de la lista
    } else {
      suma += num; // Hacemos la suma de lo que hay en "suma" más el siguiente elemento de la lista
    }
  }

  return suma; // Retornamos el valor de "suma"
}

// Creamos una función "restaOperacion()" y le psamos como prámetro una cantidad de números
double ? restaOperacion(int cantidad) {
  double ? numero; // Declaramos una variable "numero"
  double ? resta; // Declaramos una variable "resta"
  List listaNumeros = []; // Declaramos una lista vacia de números
  
  print(""); // Imprimimos un salto de línea
  
  // Haremos una iteración hasta llegar a "cantidad"
  for (int i = 0; i < cantidad; i++) {
    numero = getDouble("Digita el número ${i + 1}:"); // Pedimos un número en cada iteración
    listaNumeros.add(numero); // Agregamos el número ingresado a la lista de números
  }

  print("\nLista de números a restar: $listaNumeros\n"); // Imprimimos la lista de números

  // A continuación haremos la operación de resta de cada número de la lista
  // Recorremos todos los elementos de la lista
  for (double num in listaNumeros) {
    if (resta == null) {
      resta = num; // Si "resta" es igual a null le agregamos el primer elemento de la lista
    } else {
      resta -= num; // Hacemos la resta de lo que hay en "resta" menos el siguiente elemento de la lista
    }
  }

  return resta; // Retornamos el valor de "resta"
}

// Creamos una función "multiplicacionOperacion()" y le psamos como prámetro una cantidad de números
double ? multiplicacionOperacion(int cantidad) {
  double ? numero; // Declaramos una variable "numero"
  double ? multiplicacion; // Declaramos una variable "multiplicacion"
  List listaNumeros = []; // Pondremos una lista vacia en "listaNumeros"

  print(""); // Imprimimos un salto de línea

  // Haremos una iteración hasta llegar a "cantidad"
  for (int i = 0; i < cantidad; i++) {
    numero = getDouble("Digita el número ${i + 1}:"); // En cada iteración pediremos un número y lo enviaremos a "getDouble()"
    listaNumeros.add(numero); // Agregaremos el número ingresado a la lista de números
  }

  print("\nLista de números a multiplicar: $listaNumeros\n"); // Imprimimos la lista de números

  // A continuación haremos la operación de multiplicación de cada número de la lista
  // Recorremos todos los elementos de la lista
  for (double num in listaNumeros) {
    if (multiplicacion == null) {
      multiplicacion = num; // Si "multiplicacion" es igual a null le agregamos el primer elemento de la lista
    } else {
      multiplicacion *= num; // Hacemos la multiplicación de lo que hay en "multiplicacion" por el siguiente elemento de la lista
    }
  }

  return multiplicacion; // Retornamos el valor de "multiplicacion"
}

// Creamos una función "divisionOperacion()" y le psamos como prámetro una cantidad de números
double ? divisionOperacion(int cantidad) {
  double ? numero; // Declaramos una variable "numero"
  double ? division; // Declaramos una variable "division"
  List listaNumeros = []; // Pondremos una lista vacia en "listaNumeros"
  

  print(""); // Imprimimos un salto de línea

  // Haremos una iteración hasta llegar a "cantidad"
  for (int i = 0; i < cantidad; i++) {
    numero = getDouble("Digita el número ${i + 1}:"); // En cada iteración pediremos un número y lo enviaremos a "getDouble()"
    listaNumeros.add(numero); // Pondremos una lista vacia en "listaNumeros"
  }

  print("\nLista de números a dividir: $listaNumeros\n"); // Imprimimos la lista de números

  // A continuación haremos la operación de división de cada número de la lista
  // Recorremos todos los elementos de la lista
  for (double num in listaNumeros) {
    if (division == null) {
      division = num; // Si "division" es igual a null le agregamos el primer elemento de la lista
    } else {
      division /= num; // Hacemos la división de entre lo que hay en "division" y el siguiente elemento de la lista
    }
  }

  if (division == double.infinity) {
    return null; // Si la división retorna un valor de infinito, retornamos null
  } else {
    return division; // Si no, retornamos el valor de "division"
  }
}

// Porcentaje
double porcentajeOperacion() {
  double ? numero; // Declaramos una variable "numero"
  double ? porcentaje; // Declaramos una variable "division"
  numero = getDouble("Digita un número para calcular su porcentaje:"); // Aquí solo pediremos un número y lo enviaremos a "getDouble()"
  porcentaje = numero / 100; // Calculamos el porcentaje de "Numero"
  
  return porcentaje; // Retornamos el valor de "porcentaje"
}

// A continuación evaluaremos todo lo que el usuario ingrese por teclado
// Creamos una fución "getInt()" que recibe como parámetro una cadena de texto
int getInt(String message) {
  int data = 0; // Declaramos una variable data y la inicializamos en cero
  bool pass = false; // Esta variable nos servirá como condición para esta función

  // Mientras que pass sea igual a false, haremos lo siguiente
  while (pass == false) {
    print(message); // Imprimimos la cadena texto
    
    // Evaluamos lo que se ha ingresado por consola
    try {
      data = int.tryParse(stdin.readLineSync()!)!; // Pedimos que se ingrese un dato y lo convertimos a entero
      pass = true; // Si se ha ingresado un número, entonces "pass" será igual a true y se romperá el ciclo
    } catch (e) {
      print("Error!!!.. Debes digitar un número"); // Si se han ingresado espacios o caracteres, generamos un error hasta que se ingrese un número
    }
  }

  return data; // Retornamos el valor de "data"
}

// Creamos una fución "getDouble()" que recibe como parámetro una cadena de texto
double getDouble(String message) {
  double data = 0; // Declaramos una variable data y la inicializamos en cero
  bool pass = false; // Esta variable nos servirá como condición para esta función

  while (pass == false) {
    print(message); // Imprimimos la cadena texto
    
    // Evalumaos lo que se ha ingresado
    try {
      data = double.tryParse(stdin.readLineSync()!)!; // Pedimos que se ingrese un dato y lo convertimos a entero
      pass = true; // Si se ha ingresado un número, entonces "pass" será igual a true y se romperá el ciclo
    } catch (e) {
      print("Error!!!.. Debes digitar un número"); // Si se han ingresado espacios o caracteres, generamos un error hasta que se ingrese un número
    }
  }

  return data; // Retornamos el valor de "data"
}
